src/PostgREST/Response.hs

```diff
@@ -224,17 +225,18 @@ infoIdentResponse identifier sCache = do
           ["PATCH" | tableUpdatable table] ++
           ["DELETE" | tableDeletable table]
 
-infoProcResponse :: Routine -> Either Error.Error PgrstResponse
-infoProcResponse proc | pdVolatility proc == Volatile = respondInfo "OPTIONS,POST"
-                      | otherwise                     = respondInfo "OPTIONS,GET,HEAD,POST"
+infoProcResponse :: Routine -> Text -> Either Error.Error PgrstResponse
+infoProcResponse proc corsAllowedOrigins
+  | pdVolatility proc == Volatile = respondInfo corsAllowedOrigins "OPTIONS,POST"
+  | otherwise                     = respondInfo corsAllowedOrigins "OPTIONS,GET,HEAD,POST"
 
-infoRootResponse :: Either Error.Error PgrstResponse
-infoRootResponse = respondInfo "OPTIONS,GET,HEAD"
+infoRootResponse :: Text -> Either Error.Error PgrstResponse
+infoRootResponse corsAllowedOrigins = respondInfo corsAllowedOrigins "OPTIONS,GET,HEAD"
 
-respondInfo :: ByteString -> Either Error.Error PgrstResponse
-respondInfo allowHeader =
-  let allOrigins = ("Access-Control-Allow-Origin", "*") in
-  Right $ PgrstResponse HTTP.status200 [allOrigins, (HTTP.hAllow, allowHeader)] mempty
+respondInfo :: Text -> ByteString -> Either Error.Error PgrstResponse
+respondInfo corsAllowedOrigins allowHeader =
+  let allowedOrigins = ("Access-Control-Allow-Origin", T.encodeUtf8 corsAllowedOrigins) in
+  Right $ PgrstResponse HTTP.status200 [allowedOrigins, (HTTP.hAllow, allowHeader)] mempty
```

Comment:
To include the CORS validation for all the methods (`GET`, `POST`, etc.), the allowed origins should be handled here:

https://github.com/PostgREST/postgrest/blob/0703f27d2b0ffe60edfc6053ea5df10a3ed2c0dd/src/PostgREST/Cors.hs#L24

If you add the content of `cors-allowed-origins` to  [`corsOrigins`](https://hackage.haskell.org/package/wai-cors-0.2.7/docs/Network-Wai-Middleware-Cors.html#t:CorsResourcePolicy), then the library will verify if it's a valid Origin and return the preflight headers accordingly.

---

What I find strange is that the `OPTIONS` implementation in these lines is not done for preflight. Once the request has both `Origin` and `Access-Control-Request-Method` headers (preflight), then the `wai-cors` library will bypass this and send a default request. I couldn't find a way to make the library handle CORS requests for `OPTIONS` (not preflight, that is, only with the `Origin` header) and return by itself the headers. 

In other words, I think you'll need to handle this manually:
- Verify if there's an `Origin: <origin>` sent (otherwise leave as it is)
- Then verify if the `<origin>` is present in `cors-allowed-origins`
- If it is then include `Access-Control-Allow-Origin: <origin>` and (maybe?) `Access-Control-Allow-Credentials: true`
- If not then do not include any of those headers (CORS will fail in the browser).

If you can make the library do it, then all the better.